internal/worker/worker.go

```diff
@@ -0,0 +1,54 @@
+package worker
+
+import (
+	"context"
+	"sync"
+)
+
+type Task interface {
+	Process()
+}
+
+type WorkerPool struct {
+	taskChan    chan Task
+	wg          sync.WaitGroup
+	concurrency int
+}
+
+func NewWorkerPool(concurrency int) *WorkerPool {
+	return &WorkerPool{
+		concurrency: concurrency,
+		taskChan:    make(chan Task),
+	}
+}
+
+func (wp *WorkerPool) worker(ctx context.Context) {
+	for {
+		select {
+		case task, ok := <-wp.taskChan:
+			if !ok {
+				return
+			}
+			task.Process()
+			wp.wg.Done()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (wp *WorkerPool) Run(ctx context.Context) {
+	for i := 0; i < wp.concurrency; i++ { //nolint:intrange // requires go 1.22
+		go wp.worker(ctx)
+	}
+}
+
+func (wp *WorkerPool) AddTask(task Task) {
+	wp.wg.Add(1)
```

Comment:
The wait group should not be increased with each task: it should be increased when you create a new worker. The `WaitGroup` is required to wait for the workers to finish, not the tasks to be processed